### ISSUE TITLE
Reseting user password

### ISSUE THREAD

**Issue by Flightplas** [realm/realm-object-server#16]
Really loving the platform so far! 

Just one question. Have I missed something, or is it not possible to have a "forgot password" feature right now?

**Comment by sorenvind** [comment]
Thanks!

Our plan is to make this a new feature in the next released version. It should arrive soon ⌛️

**Comment by cliftonlabrum** [comment]
Any update on the timing of this? 

Is it coming in a matter of days, a matter of weeks, or a matter of months? :)

**Comment by jpsim** [comment]
The Realm Object Server version 1.0.0-BETA-2.2 has a CLI command `--reset-user EMAIL PASSWORD`. I'll let @sorenvind answer with any further details if necessary, or to close this issue if it's a suitable solution.

**Comment by cliftonlabrum** [comment]
If it's a CLI command, then that means I have to do it for users. I'd like to allow users to initiate their own password reset. This means making password reset available via the `SyncUser` API.

**Comment by sorenvind** [comment]
@cliftonlabrum In the current version we only have the CLI interface available. However, we'll surely consider such functionality going forward. Thanks for the great suggestion!

**Comment by TotemChris** [comment]
The way I understand it from the above (--reset-user EMAIL PASSWORD), the person issuing the CLI command would also know the user's password.  That can't be correct.  What is the right way to do this securely?  I agree with @cliftonlabrum... I can't launch without this either.  Was hoping this would be added prior to this month, but I just saw it was marked as P3 on another thread.

**Comment by cliftonlabrum** [comment]
Yeah, can you imagine the mess of a single user needing to change their password because they forgot it? An app could be live for 1 day and this will come up.

> "Hey Support, I forgot my password. I don't see anywhere to change it. Can you help?"
>
> "Yes, we can help. The only thing we can do is reset it for you. What do you want it to be?"
>
> "Uhh... you want me to *email* you my password and *tell you* what it is!?"

That's definitely not secure and very disreputable. :)

**Comment by cliftonlabrum** [comment]
Would a possible workaround be something like this?

1. **Reset Password** - Fake it on the front-end in my app and make it look like the user is "resetting" their password.
2. **New Account** - Create a new account with their existing username and set the new password they specify.
3. **Migrate Data** - Migrate their data from the previous user to the new one on the sync server.

1 and 2 would be fairly easy, but is 3 even possible?

**Comment by Montrazul** [comment]
Oh yes! I also need this!

Changing username and / or password is one of the basic requirements of each user driven application. I was really wondering why it hasnt been implemented yet.

**Comment by sorenvind** [comment]
Guys and girls, thanks for your thoughts! 

We agree with your points about the current solution: It only really works for server admins, and we need a way to solve this problem for you and your users. But we also need to get this secure and right, so we'll figure out the right solution and prioritise getting it shipped as soon as possible. Please bear with us though, it will take a little time.

**Comment by cliftonlabrum** [comment]
@sorenvind I know it's hard (impossible?) to predict software timeframes, but can we expect this in the next month or two? Or is this something that will take much longer?

**Comment by TotemChris** [comment]
Maybe allowing another 3rd party provider like OpenID or something?  I posted a feature request (Tresorit/OpenID) that may take less time to implement than baking it into Realm's native user/pass authenticator. 

In the very short term, it sounds like the best route is to go with the Apple/Google providers, and just require users to use those.

**Comment by bigfish24** [comment]
@cliftonlabrum I think the next month or two is definitely reasonable. It would help if you could provide more context on what this functionality should do. One of the most difficult aspects is how do you confirm to the user that he/she wants to reset? Typically that would involve sending an email. We would prefer not to get into the business of sending emails, since this is extremely tricky to do correctly on a server. Providing a hook to call out to an email API might be better?

The more we can learn, the faster it will be to deliver!

**Comment by bigfish24** [comment]
@TotemChris utilizing another service would be ideal. Realm Object Server's authentication can be customized and we plan to outline how to do this in the coming weeks. It would basically work just like the Google/Facebook providers we have built-in, where you would have to acquire a token from another system, pass this as custom data when logging into Realm, and then setup the Realm Object Server authentication to react to this payload and call out to the third-party's API to validate the token, triggering a Realm token to be generated.

This is the shortest path since you could utilize Auth0 or Tresorit.

**Comment by Montrazul** [comment]
Sending an email would be a basic requirement for resetting passwords. Without it it isnt really possible to verify that the real owner of the account was requesting the password reset (as long as you dont have such things like fingerprint, retina scan or something similar). 

I have written many Apps (the current App is the first app coded with realm and for now I think all my future apps will run with realm) and in every App I expect my users to register with an existing email address and I also let the users verify their email address by sending a clickable link to their email account. Only if this link got clicked the accounts are getting activated.

I see that the problem with realm is, that you dont have to use an email address as username. So if a user requests a password reset you cant verify if the user is the owner of the account. So it will be tricky.

**Comment by bigfish24** [comment]
We want to add support for profile data on the User object (to aid in doing User lookups for access control for example). So not too worried about that. The email sending is really tricky because while we could add that, it is actually quite hard to ensure the email is sent correctly. Major email providers are picky about what they label as spam if you are not careful.

**Comment by Montrazul** [comment]
That is correct but it is a very common problem and you can communicate to your users to retry the sending or to check the spam folder (even some of my google and facebook emails are sometimems moved into the spam folder).

**Comment by TotemChris** [comment]
Not to beat a dead horse, but at this point I would be ok with a user typing in their old password, then new password, and confirm password.  Just a single API to do that, with a success/fail callback.  No email.  In the situation where a user forgot, I can deal with that over email.  Pre-generate a new one for them, email it to them, and then they can reset that one once they get in.  Old school, but better than not being able to do it at all.

**Comment by tal-sh** [comment]
Hi, will this be available on the Developer edition?

**Comment by dannygrob** [comment]
I can't seem to reset a users' password. Resetting the password of an admin seems to work. I'm not sure if I'm using it correctly, but I need to stop the service, run it again with: realm-object-server --reset-user test@test.com newpassword --configuration /etc/realm/configurationfile which gives no feedback. Can't find any documentation other than this thread.

**Comment by Montrazul** [comment]
Is the following scenario possible?

1. User wants to change username and/or password
2. Server copies user related realms to temp
3. Server deletes current user
4. Server creates new user with new username and/or password
5. Server copies back realm data from temp to new user folder

Is that possible or is the realm data itself bound to the username and password and so not accessible without the original username and password?

**Comment by jahnsrud** [comment]
Any news on this?

**Comment by mklb** [comment]
I know I am not helping with this comment but I would love to see this as well. This would allow also the creation of an anonymous user BTW (which I would like to be able to do in my nearly finished App)... The flow would be something like:

1) create a user with a random username and password
2) let the user use the app without needing to sign up
3) let the user "fake" register later (we are changing the username/password here with an SKD method) so that the user does not loose its data when switching devices

**Comment by Flightplas** [comment]
Also interested if you have updates on the feature. 

It's a bit critical, because releasing a user-based app without password reseting is something I can't do.

**Comment by Montrazul** [comment]
Since it takes ages for realm to deliver the really required features I meanwhile implemented my own service to manage all that stuff (maybe for the rest of you this would also be an option). 

I implemented a small Java Spring Web Application which provides the following functionality:

- Registration
- Email Verification
- Login
- Password change
- Password reset

The flow is the following:

1. User register at the Spring Web Service with email and password
2. The web service saves original email and original password (ofcourse encrypted) and also copies it for the use of realm username and realm password
3. The web service responses success or not
4. The client now registers with the same login data to the realm mobile platform. 
5. If the realm register was successfully done send another request to the spring web service
6. Let the Web Service send the verification email

The Login and Register now always is going over my spring web service. The spring webservice is always returning the required original realm username and realm password so that the client can always authenticate at the realm mobile platform.

Changing email and password is now easy with the use of the web service. The only problem with that is..... you can actually still never change realm username and password, you just "pretend" that you are doing it. And you really need to use HTTPS, TLS 1.2, Perfect Forward Secrecy and SSL pinning... otherwise this would be too unsecure, since you always need to send the realm username and password between spring web service and client.

**Comment by caloon** [comment]
Would you mind sharing the source code of this Java Spring Web App? 

And +1 for "native" support of password reset.

**Comment by bigfish24** [comment]
We are working on this, but in the mean time you could use [Auth0](https://auth0.com/) as your custom auth provider, using our newly released custom auth functionality: https://realm.io/docs/realm-object-server/#custom-authentication

**Comment by RiverbayChris** [comment]
ROS 1.4.1 release notes has:  "Adds change password API at /auth/password endpoint."

This sounds like progress.  Any tips on how to use it?

**Comment by thevishnupradeep** [comment]
Hi,
I was looking for a way to implement password reset on Realm Mobile Platform and got here. Any update on this issue?

**Comment by bigfish24** [comment]
The current API added in 1.4.1 is an endpoint `auth/password` to change a user's own password with:

`PUT` request with body `{"token": "<TOKEN>", "password": "<NEW PASSWORD>"}`. The `<TOKEN>` is the refresh token. Currently only the Javascript SDK exposes the refresh token on the user object: https://realm.io/docs/javascript/1.2.0/api/Realm.Sync.User.html

It returns 200: `{"update": true, "accounts": <N>}`, where `<N>` is the number of accounts that were updated, it is greater than 1 if the user has multiple accounts provided by `password` auth provider.

If the token is invalid or expired, it returns
- 400: `{"type": "https://realm.io/docs/object-server/problems/invalid-credentials", "title": "The provided credentials are invalid.", "status": 400, "code": 611}`
- 403: `{"type": "https://realm.io/docs/object-server/problems/expired-refresh-token", "title": "The refresh token is expired.", "status": 403, "code": 615}`

If the token or password is not provided, it returns
400: `{"title": "Your request did not validate because of missing parameters.", "code": 602, "invalid_params": [{"name": "token", "reason": "Missing parameter 'token'!"}], "type": "https://realm.io/docs/object-server/problems/missing-parameters", "status": 400}`

We are working on a new release this week that expands on this API so you can include `user_id: <userid>` in the `PUT` body to change another user's password... however, this can only be called as an admin user (thus you need an admin's user's refresh token).

We will be adding specific APIs in the SDKs to call this vs. directly, but wanted to give an update.

**Comment by trpmap93** [comment]
The elephant in the room is the user needs the ability to reset their OWN password.  This seems to be the general point of view and the the desire of pretty much everyone in this thread.  This should be priority number one versus an administrator reset.  Don't get my vibe wrong as Realm Mobile Platform is a godsend for all of us, but fixing this huge issue would make everyone's life happier here.  The sooner the better would be the majority vote I think.

**Comment by bigfish24** [comment]
This API does allow a user to reset his/her own password, just need to pass the refresh token and new password. Our SDKs are being updated to include an API to automate this for you, but I wanted to clarify in the mean time.

**Comment by cliftonlabrum** [comment]
I was one of the first to beg for this feature, and I have since switched over to third-party auth (iCloud, Google, etc.) and have found it to be **_much_** simpler to implement and manage. 

Everyone's application needs are different, but it may be helpful to reconsider using third-party auth. You just might make your life easier. 👍 😁

**Comment by switchawk** [comment]
@bigfish24 if a user installs the app for the first time after it has been deleted (and forgot his password), he doesn't have a refresh token. In other words if I understand correctly this mechanism does not [yet] support a user who forgot his password.

**Comment by bigfish24** [comment]
Forgetting your password is different than reset, and yes that still remains to be implemented.

**Comment by switchawk** [comment]
@bigfish24 is that on the roadmap? 😬

**Comment by bigfish24** [comment]
Yes we just focused on change first. Sorry about this!

**Comment by trpmap93** [comment]
@bigfish24 that is very great of news of the SDK update to automate this.  Eagerly anticipating its arrival!
Thanks!

**Comment by trpmap93** [comment]
@cliftonlabrum thanks for the recommendation.  We have looked into this as well but with certain industries it makes it much easier (not talking technically) overall to not introduce third-parties.  Thanks again though.

**Comment by bigfish24** [comment]
Just to provide some info with regards to a forgot/reset password feature, the current thinking is:

1. Client API which `resetPasssword` triggers server creation of a token/code
2. Token/code could be returned to the client and/or passed through to a server-side event handling callback
3. The token/code can then be sent through whatever means you prefer, such as an email to the user (we don't want to support sending emails in ROS since this requires a complicated server setup to pass spam filters)
4. Once the user receives the token/code this can be used in another client API: `confirmPasswordResetWithCode:newPassword:` to change the password.

Any thoughts on this?

**Comment by switchawk** [comment]
@bigfish24 if I can easily configure a 3rd party email sending service (or my own domain), that sounds good

**Comment by riverbaymark** [comment]
@bigfish24 and @switchawk the logic and the suggestion for a 3rd part email service looks like the way to go.  Thanks guys!

**Comment by agersoncgps** [comment]
@jpsim What is the command that takes the flags --reset-user EMAIL PASSWORD?

**Comment by jpsim** [comment]
I believe the command would be `realm-object-server --reset-user EMAIL PASSWORD`, but @sorenvind should be able to confirm.

**Comment by gerbil802** [comment]
Hmm, any update on this? (for 'forgot password')

**Comment by cliftonlabrum** [comment]
@gerbil802 Yes, you can let users change their own passwords now: https://realm.io/docs/swift/latest/#changing-passwords

**Comment by gerbil802** [comment]
@cliftonlabrum this works only for logged-in users. I mean an ability to reset password for users who forgot their password and can't log in.

**Comment by ianpward** [comment]
@gerbil802 What about this admin API:
https://realm.io/docs/swift/latest/#change-passwords

**Comment by gerbil802** [comment]
@ianpward I can't see how a user can initiate that API for himself

**Comment by ianpward** [comment]
@gerbil802 I see - we hope to have something for you in our 2.0 release

**Comment by gerbil802** [comment]
Can't see it's in 2.0, I hope to hear updates about this...

**Comment by bigfish24** [comment]
@gerbil802 apologies but our documentation needs improved in this area. A user can change his/her own password. See the Obj-C docs: https://realm.io/docs/objc/3.0.1/api/Classes/RLMSyncUser.html#/c:objc(cs)RLMSyncUser(im)changePassword:completion:

Apologies that the Swift docs don't appear to list it and we need to update /docs as well.

**Comment by gerbil802** [comment]
@bigfish24 Thanks but I believe there's a misunderstanding about this ticket.  The title says _Reseting user password_  but the description actually asks for a solution for _Forgot Password_. Should I open a new ticket for that?

**Comment by bigfish24** [comment]
How would you expect reset password to work? If it involves a traditional flow like email confirmation, then this is something you will need to build yourself as Realm doesn’t handle email. Luckily you have two options: 
1. Custom authentication integration using an existing auth service such as Auth0 which offers such functionality. You can then integrate this into ROS. See https://realm.io/docs/realm_object_server/latest/
2. Extending ROS with a custom service that adds server logic to do the password reset flow using an email API service like SendGrid: https://realm.io/docs/realm_object_server/latest/

I would argue for 1 personally since that would be faster and likely provide more functionality than you could build on your own.

**Comment by gerbil802** [comment]
Thank you. The link for the (2) does not work. (I wonder what are my options for "extending" a custom service - would that server trigger the reset password for user api?)

**Comment by bigfish24** [comment]
Sorry my copy and paste with Safari on iOS 11 doesn’t actually copy the URL! Manually fixed...

**Comment by gerbil802** [comment]
@bigfish24 Thanks! I are these extensions are available on the Developer edition?

**Comment by bigfish24** [comment]
Yes the idea with ROS 2 is to open up the fact that ROS is a web framework built on Express. It has a service based architecture with several “higher level” services like Authentication and Permissions, which expose routes and you can add your own service to add new routes. The core of ROS is the SyncService which is a C++ Native module wrapped in Node layer. ROS handles the incoming websocket connections and routes it to the sync service.

**Comment by bigfish24** [comment]
Extending ROS is available in all editions, the higher paid editions allow you to scale out to more sync services, with DE you can only have 1.

**Comment by caloon** [comment]
Sorry guys, but I've been subscribed to this issue for almost a year and you repeatedly post solutions to a problem that has never been asked (namely a user resetting his own password). This thread is about a native "forgot password" solution which is a basic requirement of any professional use of realm. 

Auth0 isn't really a solution for me (and most likely many others) since it gets very expensive when scaling. So ultimately nothing has been done to solve this issue in the past year. 

I get your point that realm doesn't handle email, but why don't you build solution (2) right into realm so that we only have to set API details for sendgrid (or similar) via an object constructor or in an admin pane?

**Comment by generalpf** [comment]
I'm not seeing this work. I retrieved the admin token from `/etc/realm/admin_token.base64`, and then did this:

```
[rwalberg@revere-staging ~]$ curl -X PUT -H "Content-Type: application/json" -d '{"token":"<contents of admin_token.base64>", "password": "foobar", "user_id": "x@y.com"}' http://localhost:9080/auth/password
{"status":400,"type":"https://realm.io/docs/object-server/problems/invalid-credentials","title":"The provided credentials are invalid.","code":611}
```

Am I only supposed to use the part before or after the `:` of the contents of that file?

**Comment by generalpf** [comment]
What's the deal with this? Command-line password reset does not work either:

```
[rwalberg@revere-production ~]$ sudo realm-object-server --reset-user "<redacted>@gmail.com" "<redacted>" --configuration /etc/realm/configuration.yml
Error: History type (as specified by the Replication implementation passed to the SharedGroup constructor) was not consistent across the session
    at Error (native)
    at BaseDao (/usr/lib/node_modules/realm-object-server-developer/node_modules/ros-common/dist/dao/BaseDao.js:28:19)
    at AdminDao (/usr/lib/node_modules/realm-object-server-developer/node_modules/ros-admin-dao/dist/AdminDao.js:14:9)
    at resetUserPassword (/usr/lib/node_modules/realm-object-server-developer/.build/src/node/cli-interface.js:50:15)
    at handleCLIArguments (/usr/lib/node_modules/realm-object-server-developer/.build/src/node/cli-interface.js:156:9)
    at Object.<anonymous> (/usr/lib/node_modules/realm-object-server-developer/.build/src/node/run.js:21:46)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:389:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:504:3
```

**Comment by bigfish24** [comment]
@generalpf the correct way to do this is:
```
curl -X PUT -H "Content-Type: application/json" -H "Authorization:<user-or-admin-refresh-token>" -d '{"user_id": <user-id-to-change-password>, "data": {"new_password": <new-password>}}' http://localhost:9080/auth/password
```
The token is supplied through an `Authorization` header. The JSON payload is:
```
{
    "user_id": "5d0fbc03f4d7ab1dcc127aace224270c", // optional if user is changing his/her own password
    "data": {
        "new_password": "new-password"
    }
}
```

**Comment by bigfish24** [comment]
@caloon I am sorry about this. One of the great features of 2.0 though is that ROS is really a framework that is designed to be extended. You could add your own custom web service to perform a forgot my password flow: https://realm.io/docs/realm-object-server/latest/#adding-a-custom-service

I know this isn't a first party solution, but hopefully can help in the mean time.

**Comment by rwalbergpd** [comment]
@bigfish24 Which version of the server should this work on? If I do what you just suggested, I get:

```
{"invalid_params":[{"name":"token","reason":"Missing parameter 'token'!"}],"status":400,"type":"https://realm.io/docs/object-server/problems/missing-parameters","title":"Your request did not validate because of missing parameters.","code":602}
```

This is version 1.8.3.

The error suggests the request is _handled_ but incorrectly formatted.

**Comment by bigfish24** [comment]
@rwalbergpd This is for 2.0

**Comment by generalpf** [comment]
@bigfish24 That's all well and good but what about providing this functionality for those on version 1.x, which was current a year ago when the question was asked? It's not like upgrading to version 2.0 is a `yum update` or `apt-get upgrade` away. It's a while different platform.

In fact on May 3rd you detailed a supposed API call that would change the password, which never worked. Did it ever work? Was it built for beta and then never shipped? Genuinely confused about your release process and how I can expect to be treated if I pay for the product.

**Comment by i-schuetz** [comment]
To recap, is it possible to reset password now (with email confirmation)?

I'd expect, as some users mention above, that all this functionality is available out of the box, and we just have to enter a few settings, as with any other auth service. Most of us don't interact with the ros besides starting it and using it with the SDK in the app... it seems super inconvenient to have to dive into this just to get reset password working. 

Anyone that supports authentication should also support the full workflow for forgot/reset password. I find it a bit puzzling that it's handled so lightly in this thread, as if it were some kind of special requirement. Authentication without forgot/reset password is un-releasable. I'd immediately uninstall an app (and probably also rate it with 1 star) if I forgot my password and don't find how to reset it.

I hope there are plans to improve this, in particular now with the upcoming release of Realm Cloud.

**Comment by generalpf** [comment]
>To recap, is it possible to reset password now (with email confirmation)?

No.

>I'd expect, as some users mention above, that all this functionality is available out of the box, and we just have to enter a few settings, as with any other auth service. Most of us don't interact with the ROS besides of starting it and using it with the SDK in the app... it seems super inconvenient to have to dive into this just to get reset password working.

You're expected to use a third-party authentication provider.

>Anyone that supports authentication should also support the full workflow for forgot/reset password. I find it a bit puzzling that it's handled so lightly in this thread, as if it were some kind of special requirement. Authentication without forgot/reset password is un-releasable. I'd immediately uninstall an app (and probably also rate it with 1 star) if I forgot my password and don't find how to reset it.

100% agree.

>I hope there are plans to improve this, in particular now with the upcoming release of Realm Cloud. If you don't want to send emails then maybe make it available as an optional module, or, though not optimal, at least add a detailed tutorial to get this working to the official docs.

When we spoke to Realm's product department, they advised us the reset only works on ROS 2, and then we asked if the client SDK would be able to connect to ROS 1 and ROS 2 (since we can't force everyone to update their apps!) and were told *no*.

We switched to Firebase. It supports all of the thing you want, and the Cloud Firestore product has the same syncing capabilities as ROS. It's a shame, too, as we would have easily paid for password reset.

**Comment by bigfish24** [comment]
@i-schuetz and @generalpf First, I do apologize for the frustration and delay with this functionality. Let me try and shed some more light on the situation.

With regards to our priority previously, it is true we were more focused on custom authentication. ROS 2 was a big improvement where we redesigned the server as a framework allowing customers to add their own custom services or authentication code. Crucially, unlike ROS 1 this was directly in the Typescript codebase vs. the sidecar approach to custom auth with ROS 1. This improvement was very well received and we continued to improve rougher aspects of ROS 2 after the launch in the fall. This is what we believed was the best allocation of our resources at the time, and it does sadden me that this was a tradeoff that affected you.

With ROS 2 behind us, we have directed our focus to Realm Cloud and the new partial sync and object-level permission APIs. We knew with Realm Cloud that our existing username/password system was too limited based on all of this feedback in this thread and elsewhere. We are working to improve this situation.

First, we already started to add password reset to ROS. The initial effort was [released in 2.7.0](https://github.com/realm/realm-object-server/releases/tag/v2.7.0) where an `emailHandler` is available as a config on the `PasswordAuthProvider` to receive callbacks when a password reset is triggered. There is also an API to `updateProviderAccount` on the authentication service that can be used to trigger and complete password resets.

We haven't been able to fully document it or add this functionality into the SDKs yet but we will be doing so with Realm Cloud. As I mentioned earlier, now that we control the backend we can use an email system to reliably email a user whereas when you install it yourself, you will need to do this in the handler callback through your own means.

So please know that we are listening and trying to improve the situation. For the GA cloud launch we will have password reset functionality. However, I will caution that comparing our username/password system to other third-party providers, we will still lack some functionality and if that matters to your app we have a cloud version of custom auth through JWT. For now, we still place the majority of our focus on improving the data sync APIs.

Hopefully this helps, thank you for sharing your comments.

**Comment by joluet** [comment]
It's actually fairly easy to set up custom authentication through Firebase with the Realm Object Server. Then Firebase Auth handles all authentication related stuff (including password resets). You can even set up welcome emails. 

This is how my custom auth provider looks like:

```typescript
/// <reference path="json.d.ts"/>
import {
    BasicServer,
    auth,
    User
} from 'realm-object-server'
import {
    errors
} from 'realm-object-server'
import {
    post
} from 'superagent'
import * as admin from 'firebase-admin'
import * as serviceAccount from './firebase-key.json'

admin.initializeApp({
    credential: admin.credential.cert( < any > serviceAccount),
    databaseURL: "https://your-app.firebaseio.com"
});

export class MyCustomAuthProvider extends auth.AuthProvider {

    name = "custom/firebase"

    authenticateOrCreateUser(body: any): Promise < User > {
        return admin.auth().verifyIdToken(body.data)
            .then((decodedToken) => {
                return admin.auth().getUser(decodedToken.uid)
                    .then((userRecord) => {
                        console.log(userRecord.uid)
                        console.log(userRecord.email)
                        let userInfo : { [key: string]: string} = {};
                        console.log(userInfo)
                        return this.service.createOrUpdateUser(
                            userRecord.uid,
                            this.name, // name of the provider
                            false, // should not be an admin
                            userInfo,
                            userRecord.uid
                        )
                    })
            })
            .catch(err => {
                console.log('An error occurred on during firebase authentication: ' + err)
                return Promise.reject < User > (new errors.realm.InvalidCredentials({
                    detail: `Your credentials are invalid.`
                }))
            })
    }

    // Verifies the auth token with the Firebase auth SDK
    verifyIdToken(token) {
        return admin.auth().verifyIdToken(token)
            .then(function(decodedToken) {
                return decodedToken.uid;
            });
    }
}

```

**Comment by gerbil802** [comment]
@joluet Cool! Noob question, where does this code go?

**Comment by joluet** [comment]
You just put that file (`firebase-auth.ts`) in the `src/` directory of your realm object server.

**Comment by codegrue** [comment]
I was going to resort to using an Azure Function with an admin login into Realm Cloud to perform the password reset. Is there a better way that other's have used?

**Comment by belcast1** [comment]
@joluet thanks for your firebase suggestion.  I guess more of a typescript question, where are you storing the firebase json key in the ROS?  I am having trouble specifying the absolute path.  I have put the file in the root and trying to access with ./admin-key.json among many others .  I keep running into the module not found issue.

**Comment by joluet** [comment]
@belcast1 Yeah I had exactly the same problem. I ended up using the absolute path (/home/user/...). It's not ideal but it works.

**Comment by VagueGit** [comment]
@bigfish24 Are you able to update on how far away is password reset on Realm Cloud? We need password reset before we can release a beta of our app to our users. We are hoping for a thorough beta cycle of our app before Realm Cloud GA.

**Comment by ianpward** [comment]
@VagueGit We are putting the finishing touches on it now and we hope to have it out this week. If you would like to have a conversation on our paid cloud please reach out to info@realm.io and we can discuss your use case in depth.

**Comment by VagueGit** [comment]
Now Realm Cloud has moved to GA and still no password reset. What gives?

**Comment by nirinchev** [comment]
Password reset functionality has been rolled out across all instances now. To be able to use it, you'll need to use the user's email as their username. You can configure the sender and SMTP records in the instance settings tab or you can use the built in ones Realm Cloud provides.

**Comment by apperside** [comment]
It seem it's possibile to reset the user password. [Here](https://github.com/realm/realm-js/issues/1699) you can find the related instructions. 
Before you can make it work, you need to add some configuration to your realm object server instance following the instructions provided [here](https://docs.realm.io/platform/v/3.x/self-hosted/customize/authentication/username-password/password-reset-and-email-confirmation)
